src/cloud-api-adaptor/pkg/securecomms/sshproxy/sshproxy.go

```diff
@@ -0,0 +1,544 @@
+	} else {
+		port = int(uint64port)
+	}
+
+	if phase != "A" && phase != "K" && phase != "B" {
+		err = fmt.Errorf("illegal tag phase '%s'", phase)
+	}
+	if name == "" {
+		err = fmt.Errorf("illegal tag name '%s'", phase)
+	}
+	return
+}
+
+// NewSshPeer
+func NewSshPeer(ctx context.Context, phase string, sshConn ssh.Conn, chans <-chan ssh.NewChannel, sshReqs <-chan *ssh.Request, sid string) *SshPeer {
+	peer := &SshPeer{
+		sid:            sid,
+		phase:          phase,
+		sshConn:        sshConn,
+		ctx:            ctx,
+		done:           make(chan bool, 1),
+		outbounds:      make(map[string]*Outbound),
+		inbounds:       make(map[string]*Inbound),
+		outboundsReady: make(chan bool),
+	}
+
+	if chans == nil || sshReqs == nil {
+		logger.Fatalf("NewSshPeer with illegal parameters chans %v sshReqs %v", chans, sshReqs)
+	}
+
+	peer.wg.Add(1)
+	go func() {
+		defer peer.wg.Done()
+		for {
+			select {
+			// go ssh.DiscardRequests(sshReqs)
+			case req := <-sshReqs:
+				if req == nil {
+					peer.Close("sshReqs closed")
+					return
+				}
+				if req.WantReply {
+					if req.Type == "Phase" {
+						logger.Printf("%s phase: peer reported phase %s", phase, string(req.Payload))
+						_ = req.Reply(true, []byte(peer.phase))
+						continue
+					}
+					if phase == ATTESTATION && req.Type == "Upgrade" {
+						logger.Printf("%s phase: peer reported it is upgrading to Kubernetes phase", phase)
+						_ = req.Reply(true, []byte(peer.phase))
+						peer.upgrade = true
+						continue
+					}
+					_ = req.Reply(false, nil)
+				}
+
+			case <-ctx.Done():
+				peer.Close("Context canceled")
+				return
+			case ch := <-chans:
+				if ch == nil {
+					peer.Close("chans closed")
+					return
+				}
+				switch ch.ChannelType() {
+				default:
+					logger.Printf("%s phase: NewSshPeer rejected channel for %s", phase, ch.ChannelType())
+					_ = ch.Reject(ssh.UnknownChannelType, fmt.Sprintf("unknown channel ** type: %v", ch.ChannelType()))
+				case "tunnel":
+					name := string(ch.ExtraData())
+					<-peer.outboundsReady
+					outbound := peer.outbounds[name]
+					if outbound == nil || (outbound.Phase == "A" && phase != ATTESTATION) || (outbound.Phase == "K" && phase != KUBERNETES) {
+						logger.Printf("%s phase: NewSshPeer rejected tunnel channel: %s", phase, name)
+						_ = ch.Reject(ssh.UnknownChannelType, fmt.Sprintf("%s phase: NewSshPeer rejected tunnel channel - port not allowed: %s", phase, name))
+						continue
+					}
+					chChan, chReqs, err := ch.Accept()
+					if err != nil {
+						logger.Printf("%s phase: NewSshPeer failed to accept tunnel channel: %s", phase, err)
+						peer.Close("Accept failed")
+					}
+					logger.Printf("%s phase: NewSshPeer - peer requested a tunnel channel for %s", phase, name)
+					if outbound.Name == sshutil.KBS {
+						outbound.acceptProxy(chChan, chReqs, sid, &peer.wg)
+					} else {
+						outbound.accept(chChan, chReqs, &peer.wg)
+					}
+				}
+			}
+		}
+	}()
```

Comment:
nit: "Phase" and related strings can be put as string constant for better readability